packages/dd-trace/src/appsec/remote_config/manager.js

```diff
@@ -256,16 +250,42 @@ class RemoteConfigManager extends EventEmitter {
 
       // in case the item was already handled by kPreUpdate
       if (item.apply_state === UNACKNOWLEDGED || action === 'unapply') {
-        try {
-          // TODO: do we want to pass old and new config ?
-          const hadListeners = this.emit(item.product, action, item.file, item.id)
-
-          if (hadListeners) {
-            item.apply_state = ACKNOWLEDGED
+        const handler = this._handlers.get(item.product)
+
+        if (handler) {
+          try {
+            if (supportsAckCallback(handler)) {
+              // If the handler accepts an `ack` callback, expect that to be called and set `apply_state` accordinly
+              // TODO: do we want to pass old and new config ?
+              handler(action, item.file, item.id, (err) => {
+                if (err) {
+                  item.apply_state = ERROR
+                  item.apply_error = err.toString()
+                } else if (item.apply_state !== ERROR) {
+                  item.apply_state = ACKNOWLEDGED
+                }
+              })
+            } else {
+              // If the handler doesn't accept an `ack` callback, assume `apply_state` is `ACKNOWLEDGED`,
+              // unless it returns a promise, in which case we wait for the promise to be resolved or rejected.
+              // TODO: do we want to pass old and new config ?
+              const result = handler(action, item.file, item.id)
+              if (result instanceof Promise) {
+                result.then(
+                  () => { item.apply_state = ACKNOWLEDGED },
+                  (err) => {
+                    item.apply_state = ERROR
+                    item.apply_error = err.toString()
+                  }
+                )
+              } else {
+                item.apply_state = ACKNOWLEDGED
+              }
+            }
+          } catch (err) {
+            item.apply_state = ERROR
+            item.apply_error = err.toString()
           }
-        } catch (err) {
-          item.apply_state = ERROR
-          item.apply_error = err.toString()
         }
       }
 
```

Comment:
also (can't see the code but it's right below this line), configs will be deleted before the async handler returns (and no way to report if it was deleted correctly), is that a problem ?

Reply:
Right, but if you fail to remove it, then what? do you add it back in the config states ? or do you silently "forget" about it ?

Reply:
Conclusion after discussing this over Zoom: Yes, we could technically keep the item in the local tracer-state instead of deleting it if an error occurred when un-applying it. But that wouldn't put us in any better position: We would not be able to resolve the error automatically and try again later - we would just be in an unknown state. So it's better to give up early, log the error, and just delete the item as we do now.